server/src/main/java/org/opensearch/search/fetch/FetchPhase.java

```diff
@@ -103,101 +106,167 @@ public FetchPhase(List<FetchSubPhase> fetchSubPhases) {
-                        && docs.length >= 10) {
-                        // All the docs to fetch are adjacent but Lucene stored fields are optimized
-                        // for random access and don't optimize for sequential access - except for merging.
-                        // So we do a little hack here and pretend we're going to do merges in order to
-                        // get better sequential access.
-                        SequentialStoredFieldsLeafReader lf = (SequentialStoredFieldsLeafReader) currentReaderContext.reader();
-                        fieldReader = lf.getSequentialStoredFieldsReader()::document;
-                    } else {
-                        fieldReader = currentReaderContext.reader().storedFields()::document;
-                    }
-                    for (FetchSubPhaseProcessor processor : processors) {
-                        processor.setNextReader(currentReaderContext);
-                    }
-                }
-                assert currentReaderContext != null;
-                HitContext hit = prepareHitContext(
-                    context,
-                    fetchContext.searchLookup(),
-                    fieldsVisitor,
-                    docId,
-                    storedToRequestedFields,
-                    currentReaderContext,
-                    fieldReader
-                );
-                for (FetchSubPhaseProcessor processor : processors) {
-                    processor.process(hit);
-                }
-                hits[docs[index].index] = hit.hit();
-            } catch (Exception e) {
-                throw new FetchPhaseExecutionException(context.shardTarget(), "Error running fetch phase for doc [" + docId + "]", e);
+
+            TotalHits totalHits = context.queryResult().getTotalHits();
+            if (breakdown != null) {
+                Timer t = breakdown.getTimer(FetchTimingType.BUILD_SEARCH_HITS);
+                t.start();
+                context.fetchResult().hits(new SearchHits(hits, totalHits, context.queryResult().getMaxScore()));
+                t.stop();
+            } else {
+                context.fetchResult().hits(new SearchHits(hits, totalHits, context.queryResult().getMaxScore()));
+            }
+        } finally {
+            if (breakdown != null) {
+                context.getProfilers().getFetchProfiler().pollLastElement();
             }
         }
-        if (context.isCancelled()) {
-            throw new TaskCancelledException("cancelled task with reason: " + context.getTask().getReasonCancelled());
-        }
-
-        TotalHits totalHits = context.queryResult().getTotalHits();
-        context.fetchResult().hits(new SearchHits(hits, totalHits, context.queryResult().getMaxScore()));
-
     }
 
-    List<FetchSubPhaseProcessor> getProcessors(SearchShardTarget target, FetchContext context) {
+    List<Tuple<FetchSubPhaseProcessor, FetchSubPhase>> getProcessors(SearchShardTarget target, FetchContext context) {
```

Comment:
Do we need to use FetchSubPhase as key? Or just rely on the order in List always stay same?